wallet: Avoid underpaying transaction fees when signing taproot spends

achow101 opened the pull request:
The current taproot signing code will end up creating transactions which do not meet the feerate target. This is primarily seen when spending a taproot output with the script path. In this scenario, because the public key is known, the dummy signer would incorrectly choose to create a dummy signed transaction that used the key path spend when the actual transaction would use script path. This results in a transaction that has a significantly lower fee rate because script path spends are much larger than key path spends. This is fixed by choosing the witness with the largest stack size rather than the smallest when doing size estimation.

Unfortunately this change can result in massively overestimating the transaction fees which is undesirable. More discussion is needed to find a good solution to avoiding that problem.

There is an additional issue in the PSBT workflow where the sighash type would be appended during finalizing, but not during size estimation. That results in a transaction that is slightly larger than estimated so there is a lower feerate. This is resolved by having the DUMMY_MAXIMUM_SIGNATURE_CREATOR include the sighash type so that when funding with a watchonly wallet, we will estimate enough fee for the largest possible signature (as we do for ECDSA with high R).

Tests are updated to catch these case. `wallet_taproot.py` will now use a set feerate and check if the result meets that feerate.

achow101 committed:
Use largest stack size when estimating size

When estimating the size of a taproot witness (i.e.
DummySignatureCreator is being used), use the largest stack size rather
than the smallest. This avoids underpaying fees.

Also updates the tests to make sure that Taproot spends do not
underestimate the fee

achow101 committed:
sign: Include sighash byte for Schnorr sigs with DUMMY_MAX_SIG_CREATOR

sipa commented on src/script/sign.cpp:
```diff
@@ -570,14 +571,17 @@ class DummySignatureCreator final : public BaseSignatureCreator {
     {
         sig.assign(64, '\000');
+        if (m_include_sighash) {
+            sig.push_back((unsigned char)SIGHASH_ALL);
```
Could just do `sig.assign(64 + m_include_sighash, '\000');` in the previous line.

achow101 replied:
IMO that's a bit harder to read and I don't like the implicit bool to int conversion.

laanwj replied:
Have to agree with @achow101 here. Even though the alternative formulation is shorter ,and correct, I understand the intent of the code better this way.

sipa commented on src/script/sign.cpp:
```diff
@@ -187,6 +187,11 @@ static bool SignTaproot(const SigningProvider& provider, const BaseSignatureCrea
         sigdata.tr_spenddata.Merge(spenddata);
     }
 
+    // Detect if this is the dummy signer. If so, we want to choose the largest stack size for worst case size estimation
```
That's pretty ugly. What about adding a "bool ForFeeEstimation()" to all SignatureCreator instances, and calling that instead?

achow101 replied:
I added an `IsDummy` and changed to check that instead.

achow101 committed:
Add UseLargest function to BaseSignatureCreator

Sometimes we need to know whether we should be choosing the largest
stack size for worse case size estimation. BaseSignatureCreator can
report that to us as we only need to do this when using
DummySignatureCreator.

2310Mas commented:
Adam Ibrahim  . email samy2310@live.com .samy232075@gmail.com.

laanwj commented on src/script/sign.cpp:
```diff
@@ -570,14 +593,18 @@ class DummySignatureCreator final : public BaseSignatureCreator {
+        }
         return true;
     }
+    bool IsDummy() const override { return true; }
```
Adding a method to ask an interface if it's a certain class seems like a design inversion to me. If we can't move the actual responsibility into the class, is there some more semantic name that we could consider? I see it's assigned to `use_largest`, so maybe `UseLargest`? (then further describe in the doc-comment  `Choose the largest stack size for worst case size estimation when using this signer.`).

achow101 replied:
Changed to `UseLargest`

flack commented:
Not sure if I understand this correctly, but the underestimation could also lead to a situation where minrelaytxfee is not met, right? If so, maybe a good first step would be to prevent that.

aureleoules commented on src/script/sign.cpp:
```diff
@@ -244,25 +249,42 @@ static bool SignTaproot(const SigningProvider& provider, const BaseSignatureCrea
+                smallest_result_stack = result_stack;
+                smallest_result_stack_size = stack_size;
+            }
+            if (use_largest && (largest_result_stack.size() == 0 ||
```
nit 7f927b757748e03234d451a09309e5d788cb77d2
same above and below
```suggestion
            if (use_largest && (largest_result_stack.empty() ||
```

aureleoules reviewed (approved):
ACK 7f927b757748e03234d451a09309e5d788cb77d2 - LGTM
I also verified the test does not pass on master.

instagibbs commented on src/script/sign.cpp:
```diff
@@ -618,14 +619,17 @@ class DummySignatureCreator final : public BaseSignatureCreator {
-const BaseSignatureCreator& DUMMY_SIGNATURE_CREATOR = DummySignatureCreator(32, 32);
-const BaseSignatureCreator& DUMMY_MAXIMUM_SIGNATURE_CREATOR = DummySignatureCreator(33, 32);
+const BaseSignatureCreator& DUMMY_SIGNATURE_CREATOR = DummySignatureCreator(32, 32, false);
+const BaseSignatureCreator& DUMMY_MAXIMUM_SIGNATURE_CREATOR = DummySignatureCreator(33, 32, true);
```
annotate these args?

instagibbs commented on src/script/sign.h:
```diff
@@ -32,6 +32,9 @@ class BaseSignatureCreator {
     virtual bool CreateSchnorrSig(const SigningProvider& provider, std::vector<unsigned char>& sig, const XOnlyPubKey& pubkey, const uint256* leaf_hash, const uint256* merkle_root, SigVersion sigversion) const =0;
+
+    /** Choose the largest stack size for worst case size estimation when using this signer */
+    virtual bool UseLargest() const =0;
```
slightly philosophical question: If the wallet knows how to satisfy two script paths, will it use the more expensive one?

achow101 replied:
No. This change only affects dummy signing, but for actual signing, it will still use smallest solution.

instagibbs commented on test/functional/test_framework/util.py:
```diff
@@ -46,6 +46,13 @@ def assert_fee_amount(fee, tx_size, feerate_BTC_kvB):
+    """Assert the fee meets the feerate"""
+    target_fee = get_fee(tx_size, feerate_BTC_kvB)
+    if fee < target_fee:
+        raise AssertionError("Fee of %s BTC too low! (Should be %s BTC)" % (str(fee), str(target_fee)))
```
should be at least*

achow101 replied:
Done

aureleoules reviewed (approved):
reACK 5345b800053963eab2b4cf61662258e8821456b6

darosior reviewed (commented):
Concept ACK. But for the approach i would prefer something like https://github.com/bitcoin/bitcoin/pull/26573 instead of making the signing logic still more intricate. But that can also happen afterward.

Regarding having a more accurate estimation, i think we should eventually hint to what material in a descriptor will be available at signing time. But i'm not sure yet how. For your interest elsewhere people are testing some approaches to solve this problem: https://github.com/rust-bitcoin/rust-miniscript/pull/481.

achow101 commented:
> Concept ACK. But for the approach i would prefer something like #26573 instead of making the signing logic still more intricate. But that can also happen afterward.

The approach in this PR also has a few useful side effects. Notably, right now a PSBT that has script paths will not fill in pubkey details for the scripts. However the changes in this PR should also fix that.

FractalEncrypt commented:
> > Concept ACK. But for the approach i would prefer something like #26573 instead of making the signing logic still more intricate. But that can also happen afterward.
> 
> The approach in this PR also has a few useful side effects. Notably, right now a PSBT that has script paths will not fill in pubkey details for the scripts. However the changes in this PR should also fix that.

This is my first time commenting on a PR, so I apologize if I am breaking any protocol or etiquette here. I compiled and ran this branch and it completely solved this issue described above (PSBT script paths with missing pubkey details);

Further details;
I created a multisig address using deriveaddresses and funded it. Then I tried to create a PSBT to spend from the funded address. I was unable to create a PSBT that could be properly processed in 24.0.1. After creating the PSBT and using walletprocesspsbt in my wallet with private keys disabled, it created PSBTs missing the 'taproot_bip32_derivs' on all the multisig inputs, so my signing wallets couldn't sign the transaction. 
This is kinda crazy since I was able to create a multisig address and fund it - but now I can't spend from it. 

The code in this PR solved my issue completely. I compiled this branch and was able to easily create the PSBT with the proper input data, sign it and broadcast. Here is the signet Txid-> 72632a6c09ece682e2fe07796497ca5235280d908d027ee6900a6273e33dd2ae

I'm not knowledgeable or skilled enough to know more about this code, but as a user, I know it solved my problem. (And as a user there's no bigger problem then when you can't spend from an address that you control the keys to)

Xekyo commented on src/script/sign.cpp:
```diff
@@ -248,26 +253,43 @@ static bool SignTaproot(const SigningProvider& provider, const BaseSignatureCrea
-                GetSerializeSize(result_stack, PROTOCOL_VERSION) < GetSerializeSize(smallest_result_stack, PROTOCOL_VERSION)) {
-                smallest_result_stack = std::move(result_stack);
+            uint64_t stack_size = GetSerializeSize(result_stack, PROTOCOL_VERSION);
+            if (smallest_result_stack.empty() ||
+                stack_size < smallest_result_stack_size) {
+                smallest_result_stack = result_stack;
+                smallest_result_stack_size = stack_size;
+            }
+            if (use_largest && (largest_result_stack.empty() ||
+                stack_size > largest_result_stack_size)) {
+                largest_result_stack = result_stack;
+                largest_result_stack_size = stack_size;
```
These make sense to me, but if we correctly estimated the maximum and minimum, how would they ever get used?

achow101 replied:
I don't quite follow your question.

The estimated stack size is used to choose which result to return as multiple results could be valid. In normal signing, we use the smallest valid stack. In estimating, we use the largest.

murchandamus reviewed (commented):
Concept ACK

How would this behave in the context of changeless transactions? Would we be overestimating the target and dropping an excess to the fees? Does that get recognized when we compare coin selection solutions?

achow101 commented:
> How would this behave in the context of changeless transactions? Would we be overestimating the target and dropping an excess to the fees?

All overages due to size estimation differences are dropped to fees. The output values are set based upon the estimated size and they cannot be changed after signing. Once signing is completed and the final transaction created, if it ends up being smaller than estimated, whatever could have been saved will not be saved.

> Does that get recognized when we compare coin selection solutions?

No. Prior to actual signing, all sizes use the size estimator, which this PR changes to use the worst case scenario.

achow101 commented:
Superseded by #26573
